### DISCUSSION TITLE
OpenAPI callbacks & webhooks + Sync-to-Async triggers features and design

### DISCUSSION THREAD

**Discussion by lbroudoux** [microcks/microcks#1039] [Ideas]
Hey Microcks community!

As we're starting a new year and a new development cycle, I'd like to share and discuss the design ideas I have around 3 related topics:
* OpenAPI callbacks support (as introduced in OAS 3.0),
* OpenAPI webhooks support (as introduced in OAS 3.1),
* Sync-to-Async triggers (as introduced as a personal idea 😉 )

These elements were on our roadmap for a long time (https://github.com/orgs/microcks/projects/1?pane=issue&itemId=26464135), and I saw some demands recently on those topics. As these topics are closely related, I spent some time thinking about the best way to add support and implementation in Microcks. 

Please consider this discussion as an opportunity to comment and share what you'd like to see and your feelings about the proposed plan. Feel free to react to anything and propose an alternative approach if it sounds better to you!

I'll expose here my analysis on 3 different topics and then different options I have considered for representation and implementation.

**References:**
* Callbacks (OAS 3.0) intro: [Callbacks](https://swagger.io/docs/specification/callbacks/)
* Callbacks spec: [OpenAPI Specification v3.1.0 | Introduction, Definitions, & More](https://spec.openapis.org/oas/latest.html#callbackObject)
* Webhooks (OAS 3.1) example: [OpenAPI-Specification/examples/v3.1/webhook-example.yaml at main · OAI/OpenAPI-Specification](https://github.com/OAI/OpenAPI-Specification/blob/main/examples/v3.1/webhook-example.yaml)
* Bump on Webhooks: [Documenting your OpenAPI webhooks · Bump.sh](https://bump.sh/blog/documenting-your-OpenAPI-webhooks/)
* Microcks Sync-to-Async triggers: [Sync to Async triggers · microcks · Discussion \#694](https://github.com/orgs/microcks/discussions/694)
* Webhooks support: [When can we have a support for webhooks? · microcks · Discussion \#915](https://github.com/orgs/microcks/discussions/915)

## Analysis
Callbacks are defined with both request (from API owner to user) and response (from user server to API owner)
> Is the response useful in the context of mocking? I don’t think so…

Callbacks unsubscribe method is not defined by the spec. It’s up to the implementation to provide an `/unsubscribe` operation and a subscription identification mechanism (based on id or whatever)
> I think that we can ignore this point in Microcks and consider we only have ephemeral subscriptions that are consecutive to the `/subscribe` operation invocation and are not persisted. IMHO this is the sweet spot for OpenAPI callbacks vs AsyncAPI: callbacks are best suited for « events consecutive to an operation one on specific one entity » ; AsyncAPI is well suited for « stream of events not necessarily related to an operation » (this last one tends to be mitigated with « request-response » style integrated into AsyncAPI v3.0 though…)

Webhooks - even if closely related to the callbacks feature - describe requests initiated other than by an API call, for example, by an out-of-band registration. We can typically think here about the webhooks that can be registered via the GitHub administration console and allow third-party apps to receive HTTP notification for events (whatever the entity)
> Webhooks sound much more like AsyncAPI event publication to me. 

It's a very common pattern to have a synchronous interaction (ie. invocation of Rest/Soap/Graph/Grpc API operation) triggering the publication of an asynchronous event on a channel (and thus an AsyncAPI subscribe operation). So why not be able to define this triggering from the mocking step in API lifecycle?
> I see 2 different areas to address, which could be easy to set up in Microcks. The first one is about how to define this link using API specs and in Microcks ; the second one is about how to propagate context from sync request to async message to get coherent content. 

## Example cases
These are simplified representations with some levels of complexity ignored to make things simple to understand.
### Callback
```yaml
/subscribe
  post:
    requestBody:
      examples:
        EXAMPLE_1: {}
    response:
      examples:
        EXAMPLE_1: {}
    callbacks:
      createEvent:
        '{$request.body#/callbackUrl}':
          requestBody:
            examples:
              EXAMPLE1: {}
          responses:
            examples:
              EXAMPLE1: {}   # May actually be ignored.
      updateEvent:
        '{$request.body#/callbackUrl}':
          requestBody:
            examples:
              EXAMPLE1: {}
          responses:
            examples:
              EXAMPLE1: {}   # May actually be ignored.
```

### Webhook
```yaml
webhooks:
  # Each webhook needs a name
  newPet:
    post:
      requestBody:
        description: Information about a new pet in the system
        examples:
          EXAMPLE_1: {}
      responses:
        examples:
          EXAMPLE_1: {}	# May actually be ignored.
```

### Sync-to-Async triggers
```yaml
openapi: 3.1
info:
  title: User registration API
  version: 1.0.0
paths:
  /user:
    post:
      requestBody:
        examples:
          EXAMPLE_1: {"name": "John Doe"}
      response:
        examples:
          EXAMPLE_1: {"status": "registered"}
      x-microcks:
        trigger: 'User signed-up API:0.0.1:PUBLISH /user/signups'	# Exclusive with `triggeredBy` in AAS
```

```yaml
asyncapi: 2.6
info:
  title: User signed-up API
  version: 1.0.0
channels:
  /user/signups:
    subscribe:
      message:
        examples:
          EXAMPLE_1: {"id": "{{ uuid() }}", "name": "{{ request.body/name }}"}
          EXAMPLE_2: {}
      x-microcks:
        triggeredBy: 'User registration API:1.0.0:POST /user'		# Exclusive with `trigger` in OAS
```

### Option 1 for callbacks
Callbacks are defined as other pairs of `Request`/`Response` attached to the same operation and related to the origin example.

In that case, we need a new discriminator (or type of request and response) to tell Microcks it’s a callback-related request or response.
Operation `POST /subscribe` => 1 Request and 1 Response with `operationId` pointing to operation => 2 Requests (and 2 Responses) of type `Callback` with the reference of the event in the Spec + the endpoint expression (`{$request.body#/callbackUrl}`)

### Option 2 for callbacks
Callbacks are defined as additional  `EventMessage`s attached to the same operation and related to the origin example.

`EventMessage` already holds a `dispatchCriteria` that can be used to hold the endpoint expression  (`{$request.body#/callbackUrl}`). We would need an additional property to keep track of event name (`createEvent` or `updateEvent`) for the sake of completeness.

### Option 1 for webhooks
We can ignore them and simply argue that people have to consider AsyncAPI instead 😉

### Option 2 for webhooks
Webhooks are defined similarly to AsyncAPI using `EventMessage`s attached to an operation bound to a `REST API` in Microcks. In that case, the REST API may have a traditional HTTP-Verb operation + `SUBSCRIBE`like operation in AsyncAPI (meaning that the consumer will receive events)

### Options for Sync-to-Async triggers
Mapping onto Microck’s internal model is obvious as we have 2 different APIs with already existing `Request`/`Response`/`EventMessage` types that can live independently. One thing we should add though is this concept of `trigger`:
1) As an independent association table (or collection in the DocumentDB world)
2) As an additional property on the operation definition. 

Supporting both side’s specifications - `trigger` and `triggerBy` - can be tricky, and it may lead to inconsistencies. Things would be easier just supporting the OpenAPI side of things with `trigger` (it would also not trigger an update of unrelated API). The 2nd option seems more natural as it’s where the business logic of event triggering should be located.

> Question: should we allow multi-values for `trigger` property? (Aka triggering multiple events from the same REST operation invocation). I would think so, but I would need confirmation.

### Option 1 for callbacks
Implement sending back requests in the main web app component. We may use Spring internal `ApplicationEvents` to have temporal decoupling with the main mock invocation thread (like we’re already doing for `metrics` gathering or event publication to async-minion). We need a way to wait for a specified (short-)time before sending the event without locking threads or resources (see concern `#2` below) 

It fits whatever option is used for modeling.

### Option 2 for callbacks
Implement sending back requests in the async-minion component. It fits whatever option is used for modeling, but it adds a hard dependency on async stuff, even for « simple » OpenAPI use cases. Here again, we need a way to wait for a specified (short-)time before sending the event without locking threads or resources (see concern `#2` below). The scheduler we used on Quarkus is much more appropriate for long-running tasks, not for ephemeral and short-lived notifications.

### Option 1 for webhooks
Implement sending back requests in the main web app component. As we just have a basic scheduler on the main component (just for import jobs), this would require duplicating the scheduling and long-running task logic we already have in the async-minion… 

### Option 2 for webhooks
Implement sending back requests in the async-minion component.

This option fits better with the modeling option 2 for webhooks as we can solely rely on the `EventMessage` definitions to send stream of messages. The process would be as follows:
* At startup: gather all candidate AsyncAPI (as of today) + gather all webhooks `EventMessage` definitions
* Depending on configured frequencies (see concern `#4` below), attach AsyncAPI and webhooks via the same `AsyncMockDefinition`
* One different thing would be how we retrieve the target endpoints (see concern `#3` below):
  * Endpoints are static configuration in the case of AsyncAPI
  * Endpoints may be dynamically registered/un-registered in the case of REST webhooks

### Options for Sync-to-Async triggers
Implementation of sending triggered event must be realized in the async-minion component. Triggering command will be materialized as an event sent from the Microcks main component to the async-minion (via Kafka or WebSocket for the uber images) and message reception (filtered by minion-supported bindings) will trigger a single message publication.  `ProducerManager` will need refactoring to allow the reuse of single message publication outside of the scheduled jobs.

How do we reuse content from the incoming request into our triggered event messages? See concern `#1` below.

## Concerns

1) We need a way to transfer context from the original request to the callback of the triggered message payload. Template expressions (`{{request.body/path}}`) seem to fit well here. In the context of AsyncAPI triggering, we may use the presence of `{{ (request|response).* }} `expression as a filter to know that we should use this as a template for sending messages. These same messages would be ignored for regular AsyncAPI mock events publication.
2) We need a way to specify/know the delay between original request processing and callback submission. This can be a low random number (2-5 secs) in case of a single callback events but we need explicit information if there is more than 1 callback - in order to keep the temporal logic of events (`createEvent` must be sent back before `updateEvent` but Microcks can’t guess this based on event names).
3) Even if we model Webhooks following the option 2, we’d need a way to realize the out-of-band registration of target endpoints. Contrary to AsyncAPI brokers that can be typically shared by consumers (let’s say - most of the time), webhooks endpoints tend to be multiples (you know, the web nature…). We may need to provide a Microcks based API to allow registration/unregistration of endpoints for webhooks for an API and version.
4) We need a way to specify/know the frequency for webhook events publication. We can follow here the same logic as for AsyncAPI: default frequency or the one specified via a `x-microcks/frequency` property at the operation/webhook level.

## Summary

Proposed priorities:
1) OAS Callbacks implementation (based on community requests)
2) Sync-to-Async triggers (based on experience and to ease implementation of webhooks later on)
3) OAS Webhooks implementation

Proposed options:
* OAS Callback: **Option 2** regarding model representation + **Option 1** regarding runtime implementation
* Sync-to-Async triggers: `trigger` custom property support only, as a multi-valued property 
* OAS Webhooks: **Option 2** regarding model representation + **Option 2** regarding runtime implementation

**Comment by vinodbalasu** [comment]
Most fintech companies widely adopt the Sync-to-Async trigger pattern, where third-party APIs provide an immediate acknowledgment response upon receiving a request, but the actual processing is completed asynchronously. The final status or result is communicated later via a callback to a webhook endpoint.

This pattern is especially common in the payments and transaction processing domains, where real-time acknowledgments are crucial for user experience, but backend processing may involve delays due to settlement, validation, or third-party dependencies.

📌 Example:
A client initiates a payment request to a third-party payment gateway.

The payment gateway immediately responds with:


```
{
  "status": "RECEIVED",
  "paymentId": "12345"
}
```

After processing (which could take seconds or minutes), the gateway sends an HTTP POST callback to the client’s webhook:

```
{
  "paymentId": "12345",
  "status": "SUCCESS",
  "processedAt": "2025-06-17T12:00:00Z"
}
```
This asynchronous update mechanism is a critical requirement across the fintech industry and is heavily used in UPI, BBPS, card processing, loan disbursals, and KYC verification workflows.
@lbroudoux @yada Can we please include this in the road map.

**Reply by lbroudoux** [reply]
Hi @vinodbalasu 

I understand that the **OpenAPI callbacks** implementation is the one you're looking for in priority. Is it correct?
Have you been through the different options I described above regarding the implementation? Any thoughts and comments on them?

Cheers,

**Reply by vinodbalasu** [reply]
Yes, supporting OpenAPI Callbacks (OAS 3.0) would be super valuable, especially for mocking real-world Sync-to-Async workflows where an initial acknowledgment is sent, and callbacks are later triggered to our webhook. This is crucial for simulating complex payment or onboarding flows during testing.

Similarly, OpenAPI Webhooks (OAS 3.1) would be a great addition to mock event-driven APIs, particularly when our system is the passive receiver.

But what I would really love to see is support for x-microcks: trigger. Having this as an extra layer would give us fine control over when and how callbacks are fired during mocks, making the simulation far more flexible and realistic. This would significantly improve the ability to test end-to-end async interactions in controlled, predictable ways before moving to real environments.

**Reply by lbroudoux** [reply]
Ok thanks for the confirmation.

> what I would really love to see is support for x-microcks: trigger. 

From what I described above, this one would only target the OpenAPI + AsyncAPI scenario and not the one with callbacks or webhooks where interpretation of existing elements at the OpenAPI spec level is enough. 

So to sum-up, could we say that your priority would be:
1. **OpenAPI callbacks**
2. **OpenAPI call trigger AsyncAPI publication**

Is that correct? 

BTW, would you like to champion the implementation, test and documentation of these features? By "championing" I mean being an active tester, reviewer and commenter on what we deliver. It's important for us to have champions so that we have a better confidence on what we deliver actually worths it and it well tested.

**Comment by vinodbalasu** [comment]
Yes, my top priority is supporting OpenAPI Callbacks, as that directly enables mocking the Sync-to-Async flows we rely on in payment and onboarding scenarios — where we receive an immediate acknowledgment and later get asynchronous status updates via webhooks.

Also, support for x-microcks: trigger would be super valuable, as it would give us fine control over when and how callbacks are fired during mocks, making the simulations much closer to real-world scenarios.

The OpenAPI triggering AsyncAPI publication would be great to have down the line, but for now, my primary focus is on the Callbacks pattern.

I’d be happy to champion this — I can actively test, review, and provide feedback to ensure the feature is robust and solves real-world problems.

**Reply by lbroudoux** [reply]
Awesome! Let me check if/how we can prioritize this for a future release. I expect to find some free cycles during summer (July?) to work on this.

**Comment by lbroudoux** [comment]
Hey there! I'm back to tackling these topics thanks to the impetus from #1926 and the proximity of some topics.

While scratching my head around this, I review and modify some of the options I previously selected. Let me sum up my thoughts.

I previously proposed options for model representation and runtime implementation that I think didn't fully align with the Microcks guidelines 'til now. So, for example, when I proposed using `EventMessage` to represent Callbacks, it was in fact a shortcut omitting a part of the model and forgetting the fact that Microcks is also about communication (displaying complete exchange instead of just what we have interest when mocking) and was also about testing (and for that we need both request and response representation for callbacks).

So I review the options and think in terms of embracing the variety of features and use cases we have to support.

Here's the classes diagram I can imagine at the moment, to support all our use cases:

 ```mermaid
classDiagram
    class ReplyInfo {
        +String address
        +String addressLocation
    }
    class Exchange 
    <<abstract>> Exchange
    Service *-- Operation : composition
    Operation *-- Exchange : composition
    class Operation {
        +ReplyInfo replyInfo
    } 
    Exchange <|-- RequestResponsePair
    Exchange <|-- UnidirectionalEvent
    Exchange <|-- RequestReplyEvents
    RequestResponsePair <|-- RequestResponsePairWithCB
    class RequestResponsePair {
      +Request request
      +Response response  
    }
    class UnidirectionalEvent {
      +EventMessage eventMessage
    }
    class RequestReplyEvents {
      +EventMessage request
      +EventMessage reply  
    }
    class RequestResponsePairWithCB {
      +[RequestResponsePair] callbacks 
    }
    class Message 
    <<abstract>> Message
    Message <|-- Request
    Message <|-- Response
    Message <|-- EventMessage
    class Message {
       +String callbackId 
    }
    class Request {
       +String responseId
    }
    class Response {
       +String dispatchCriteria 
    }
    class EventMessage {
       +String dispatchCriteria
       +String replyId
    }
```

For OpenAPI callbacks, I imagine introducing the new `RequestResponsePairWithCB` exchange type to allow integrating callbacks (that are just request/response pairs after all). We'll also need to keep track of the callback identifier (is this a `createEvent ' or an `updateEvent` callback?) - so I think of introducing a new `callbackId` at the `Message` level. I don't like that idea much, and I would prefer something more generic, but naming is hard!

> [!NOTE]
>  In the case of callbacks, we also need some kind of address location (typically something like `{$request.body#/callbackUrl}`) to be able to find the callback URL from original incoming request. I think I could reuse the `ReplyInfo` operation for this. What do you think?

I am still a bit stuck on the callback-specific question of how to specify/know the delay between the original request processing and the callback submission. This can be a low random number (2-5 secs) in case of a single callback events but we need explicit information if there is more than 1 callback - in order to keep the temporal logic of events (`createEvent` must be sent back before `updateEvent` but Microcks can’t guess this based on event names). but that's another story...

For OpenAPI webhooks, I realize that there was, in fact, an additional `Operation` that also follows the request/response paradigm. The only difference is that the registration/unregistration endpoints are out of scope in the spec. 

What do you think? Does it make sense?
